File: neet_code_150_cpp/sliding_window/longest_substring_no_repeat.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <deque>
#include <algorithm>

class Solution {
    public:
        // AC
        // O(N^2)で遅い
        int lengthOfLongestSubstringSq(std::string s) {
            if(!s.size()) {
                return 0;
            }

            std::unordered_set<char> set;
            std::vector<int> len_list;
            int len_tmp = 1;

            for(int i = 0; i < s.size(); ++i) {
                set.insert(s[i]);
                for(int j = i + 1; j < s.size(); ++j) {
                    if(set.find(s[j]) == set.end()) {
                        set.insert(s[j]);
                        ++len_tmp;
                    } else {
                        break;
                    }
                }

                len_list.push_back(len_tmp);
                set.clear();
                len_tmp = 1;
            }

            return *std::max_element(len_list.begin(), len_list.end());
        }

};

// Rustの模範解答より
class SolutionAnsRust {
    public:
        // AC
        // O(N^2)
        int lengthOfLongestSubstring(std::string s) {
            std::deque<char> set;
            int result = 0;

            auto is_contain = [&](char c) {
                for(auto& c_dq: set) {
                    if(c == c_dq) {
                        return true;
                    }
                }

                return false;
            };

            for(auto& c : s) {
                while(is_contain(c)) {
                    set.pop_front();
                }

                set.push_back(c);
                result = std::max(result, (int)set.size());
            }

            return result;
        }
};

// 模範解答
class SolutionAns {
    public:
        int lengthOfLongestSubstring(std::string& s) {
            std::unordered_set<char> chars;
            int maxSize = 0;
            int i = 0, j = 0;
            while (j < s.size()){
                while (chars.find(s[j]) != chars.end()){
                    chars.erase(s[i]);
                    ++i;
                }
                maxSize = std::max(maxSize, j - i + 1);
                chars.insert(s[j]);
                ++j;
            }
            return maxSize;
        }
};

int main(void) {
    std::string case_1 = "abcabcbb";
    std::string case_2 = "bbbbb";
    std::string case_3 = "pwwkew";
    std::string case_4 = "xyz";
    std::string case_5 = "au";
    std::string case_6 = "a";
    std::string case_7 = "dvdf";

    Solution s_1;

    std::cout << "case_1 " << s_1.lengthOfLongestSubstringSq(case_1) << std::endl;
    std::cout << "case_2 " << s_1.lengthOfLongestSubstringSq(case_2) << std::endl;
    std::cout << "case_3 " << s_1.lengthOfLongestSubstringSq(case_3) << std::endl;
    std::cout << "case_4 " << s_1.lengthOfLongestSubstringSq(case_4) << std::endl;
    std::cout << "case_5 " << s_1.lengthOfLongestSubstringSq(case_5) << std::endl;
    std::cout << "case_6 " << s_1.lengthOfLongestSubstringSq(case_6) << std::endl;
    std::cout << "case_7 " << s_1.lengthOfLongestSubstringSq(case_7) << std::endl;

    SolutionAnsRust s_ans_rs;

    std::cout << "case_1 " << s_ans_rs.lengthOfLongestSubstring(case_1) << std::endl;
    std::cout << "case_2 " << s_ans_rs.lengthOfLongestSubstring(case_2) << std::endl;
    std::cout << "case_3 " << s_ans_rs.lengthOfLongestSubstring(case_3) << std::endl;
    std::cout << "case_4 " << s_ans_rs.lengthOfLongestSubstring(case_4) << std::endl;
    std::cout << "case_5 " << s_ans_rs.lengthOfLongestSubstring(case_5) << std::endl;
    std::cout << "case_6 " << s_ans_rs.lengthOfLongestSubstring(case_6) << std::endl;
    std::cout << "case_7 " << s_ans_rs.lengthOfLongestSubstring(case_7) << std::endl;

    SolutionAns s_ans;
    std::cout << "case_1 " << s_ans.lengthOfLongestSubstring(case_1) << std::endl;
    std::cout << "case_2 " << s_ans.lengthOfLongestSubstring(case_2) << std::endl;
    std::cout << "case_3 " << s_ans.lengthOfLongestSubstring(case_3) << std::endl;
    std::cout << "case_4 " << s_ans.lengthOfLongestSubstring(case_4) << std::endl;
    std::cout << "case_5 " << s_ans.lengthOfLongestSubstring(case_5) << std::endl;
    std::cout << "case_6 " << s_ans.lengthOfLongestSubstring(case_6) << std::endl;
    std::cout << "case_7 " << s_ans.lengthOfLongestSubstring(case_7) << std::endl;
}

File: neet_code_150/src/sliding_window/longest_substring_no_repeat.rs
use std::collections::{HashSet, VecDeque};

struct Solution {}
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        if s.len() == 0 {
            return 0;
        }
        let mut set = HashSet::new();
        let mut len_list = vec![];
        let mut len_tmp = 0;
        for c in s.chars() {
            if !set.contains(&c) {
                set.insert(c);
                len_tmp += 1;
            } else {
                len_list.push(len_tmp);
                len_tmp = 1;
                set.clear();
                set.insert(c);
            }
        }
        len_list.push(len_tmp);

        println!("len_list: {:?}", len_list);
        *len_list.iter().max().unwrap()
    }

    // AC : O(N^2)
    pub fn length_of_longest_substring_sq(s: String) -> i32 {
        if s.len() == 0 {
            return 0;
        }
        let s_vec: Vec<char> = s.chars().collect();

        let mut set = HashSet::new();
        let mut len_list = vec![];
        let mut len_tmp = 1;

        for i in 0..s_vec.len() {
            set.insert(s_vec[i]);
            for j in (i + 1)..s_vec.len() {
                if !set.contains(&s_vec[j]) {
                    set.insert(s_vec[j]);
                    len_tmp += 1;
                } else {
                    break;
                }
            }
            len_list.push(len_tmp);
            set.clear();
            len_tmp = 1;
        }

        println!("len_list: {:?}", len_list);
        *len_list.iter().max().unwrap()
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut set: VecDeque<char> = VecDeque::new();
        let mut longest = 0;

        for c in s.chars() {
            while set.contains(&c) {
                set.pop_front();
            }

            set.push_back(c);
            longest = longest.max(set.len());
        }

        longest as i32
    }
}

// AC
// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let vec_char:Vec<char> = s.chars().collect();
        let mut set:HashSet<char> = HashSet::new();
        let mut result = 0;
        let (mut i, mut j) = (0, 0);

        while j < vec_char.len() {
            while set.contains(&vec_char[j]) {
                set.remove(&vec_char[i]);
                i += 1;
            }

            result = result.max(j - i + 1);
            set.insert(vec_char[j]);
            j += 1;
        }

        result as i32
    }
}

fn main() {
    let case_1 = "abcabcbb".to_string();
    let case_2 = "bbbbb".to_string();
    let case_3 = "pwwkew".to_string();
    let case_4 = "xyz".to_string();
    let case_5 = "au".to_string();
    let case_6 = "a".to_string();
    let case_7 = "dvdf".to_string();

    println!("lenght_of_longest_substring");
    println!(
        "case_1: {}",
        Solution::length_of_longest_substring(case_1.clone())
    );
    println!(
        "case_2: {}",
        Solution::length_of_longest_substring(case_2.clone())
    );
    println!(
        "case_3: {}",
        Solution::length_of_longest_substring(case_3.clone())
    );
    println!(
        "case_4: {}",
        Solution::length_of_longest_substring(case_4.clone())
    );
    println!(
        "case_5: {}",
        Solution::length_of_longest_substring(case_5.clone())
    );
    println!(
        "case_6: {}",
        Solution::length_of_longest_substring(case_6.clone())
    );
    println!(
        "case_7: {}",
        Solution::length_of_longest_substring(case_7.clone())
    );

    println!("lenght_of_longest_substring_sq");
    println!(
        "case_1: {}",
        Solution::length_of_longest_substring_sq(case_1.clone())
    );
    println!(
        "case_2: {}",
        Solution::length_of_longest_substring_sq(case_2.clone())
    );
    println!(
        "case_3: {}",
        Solution::length_of_longest_substring_sq(case_3.clone())
    );
    println!(
        "case_4: {}",
        Solution::length_of_longest_substring_sq(case_4.clone())
    );
    println!(
        "case_5: {}",
        Solution::length_of_longest_substring_sq(case_5.clone())
    );
    println!(
        "case_6: {}",
        Solution::length_of_longest_substring_sq(case_6.clone())
    );
    println!(
        "case_7: {}",
        Solution::length_of_longest_substring_sq(case_7.clone())
    );

    println!("模範解答");
    println!(
        "case_1: {}",
        SolutionAns::length_of_longest_substring(case_1.clone())
    );
    println!(
        "case_2: {}",
        SolutionAns::length_of_longest_substring(case_2.clone())
    );
    println!(
        "case_3: {}",
        SolutionAns::length_of_longest_substring(case_3.clone())
    );
    println!(
        "case_4: {}",
        SolutionAns::length_of_longest_substring(case_4.clone())
    );
    println!(
        "case_5: {}",
        SolutionAns::length_of_longest_substring(case_5.clone())
    );
    println!(
        "case_6: {}",
        SolutionAns::length_of_longest_substring(case_6.clone())
    );
    println!(
        "case_7: {}",
        SolutionAns::length_of_longest_substring(case_7.clone())
    );

    println!("C++の模範解答より");
    println!(
        "case_1: {}",
        SolutionAnsCpp::length_of_longest_substring(case_1.clone())
    );
    println!(
        "case_2: {}",
        SolutionAnsCpp::length_of_longest_substring(case_2.clone())
    );
    println!(
        "case_3: {}",
        SolutionAnsCpp::length_of_longest_substring(case_3.clone())
    );
    println!(
        "case_4: {}",
        SolutionAnsCpp::length_of_longest_substring(case_4.clone())
    );
    println!(
        "case_5: {}",
        SolutionAnsCpp::length_of_longest_substring(case_5.clone())
    );
    println!(
        "case_6: {}",
        SolutionAnsCpp::length_of_longest_substring(case_6.clone())
    );
    println!(
        "case_7: {}",
        SolutionAnsCpp::length_of_longest_substring(case_7.clone())
    );
}
